MapboxNavigation/NavigationMapView.swift

```diff
@@ -709,6 +720,190 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
     }
 
     // MARK: - Vanishing route line methods
+
+    var primaryRoutePoints: RoutePoints?
+    var primaryRouteLineGranularDistances: RouteLineGranularDistances?
+    var primaryRouteRemainingDistancesIndex: Int?
+    var routeLineTracksTraversal: Bool = false
+    private var startDate = Date()
+    public var fractionTraveled: Double = 0.0
+    var preFractionTraveled: Double = 0.0
+    var timer: Timer? = nil
+    
+    struct RoutePoints {
+        var nestedList: [[[CLLocationCoordinate2D]]]
+        var flatList: [CLLocationCoordinate2D]
+    }
+    
+    struct RouteLineGranularDistances {
+        var distance: Double
+        var distanceArray: [RouteLineDistancesIndex]
+    }
+    
+    struct RouteLineDistancesIndex {
+        var point: CLLocationCoordinate2D
+        var distanceRemaining: Double
+    }
+    
+    func initPrimaryRoutePoints(route: Route) {
+        primaryRoutePoints = parseRoutePoints(route: route)
+        primaryRouteLineGranularDistances = calculateRouteGranularDistances(coordinates: primaryRoutePoints?.flatList ?? [])
+    }
+    
+    /**
+    Tranform the route data into nested arrays of legs -> steps -> coordinates.
+    The first and last point of adjacent steps overlap and are duplicated.
+    */
+    func parseRoutePoints(route: Route) -> RoutePoints {
+        let nestedList = route.legs.map { (routeLeg: RouteLeg) -> [[CLLocationCoordinate2D]] in
+            return routeLeg.steps.map { (routeStep: RouteStep) -> [CLLocationCoordinate2D] in
+                if let routeShape = routeStep.shape {
+                    if !routeShape.coordinates.isEmpty {
+                        return routeShape.coordinates
+                    } else { return [] }
+                } else {
+                    return []
+                }
+            }
+        }
+        let flatList = nestedList.flatMap { $0.flatMap { $0.compactMap { $0 } } }
+        return RoutePoints(nestedList: nestedList, flatList: flatList)
+    }
+    
+    /**
+    Find and cache the index of the upcoming [RouteLineDistancesIndex].
+    */
+    public func updateUpcomingRoutePointIndex(routeProgress: RouteProgress) {
+        let currentLegProgress = routeProgress.currentLegProgress
+        let currentStepProgress = routeProgress.currentLegProgress.currentStepProgress
+        if primaryRoutePoints != nil {
```

Comment:
Please return early here by using `guard let` statement.

Reply:
Thanks for the suggestion, I'll improve this.